### ISSUE TITLE
THLapackCheck on windows

### ISSUE THREAD

**Issue by aiworld** [torch/torch7#387]
Got this error when installing on windows FYI:

`c:\users\a\appdata\local\temp\luarocks_torch-scm-2-2919\torch7\lib\th\generic/TH TensorLapack.c(210) : error C2059: syntax error : ')'`

I believe this is a variadic macro issue, but it was easily fixed with the following:

https://github.com/aiworld/torch7/commit/4fe36336f918c235efaff98ef7469e248b4fce49#diff-6232529b35671ad635fed629c58a97f8

Everything else built fine for me on windows! Wooooo :rocket:

**Comment by hughperkins** [comment]
Nice :-)  How did you build it?  mingw? cygwin?  something else?

**Comment by aiworld** [comment]
Thanks. The hard work was definitely done by @diz-vara's amazing https://github.com/diz-vara/luajit-rocks, with some modifications like getting the lastest version of torch from here for dependencies on functions like `THAtomicAddLong`.

**Comment by hughperkins** [comment]
Ah, nice.  So, doesnt need cygwin/mingw in fact?  Can compile using msvc, eg via SDK?  Thats cool :-)

**Comment by aiworld** [comment]
Yep! Through cmake / nmake. I highly recommend jom for quick, parallelized replacement for nmake as well. https://wiki.qt.io/Jom Just rename in `jom.exe` to `nmake.exe` and to build with all your cores :smile: .

**Comment by nathanprat** [comment]
I can confirm this error is still happening, and your solution works.
Could you make a pull request?

I guess it is related to the way MSVC expends `__VA_ARGS__` in `lib\TH\THLapack.h` (where `THLapackCheck` is defined). Read [also](https://msdn.microsoft.com/en-us/library/ms177415(v=vs.120).aspx)
I tried to define it this way :
```
#ifndef _WIN32
	#define THLapackCheck(fmt, func, info , ...)						\
	if (info < 0) {														\
	  THError("Lapack Error in %s : Illegal Argument %d", func, -info); \
	} else if(info > 0) {												\
	  THError(fmt, func, info, ##__VA_ARGS__);							\
	}
#else
	#define THLapackCheck(fmt, func, info , ...)						\
	if (info < 0) {														\
	  THError("Lapack Error in %s : Illegal Argument %d", func, -info); \
	} else if(info > 0) {												\
	  THError(fmt, func, info, __VA_ARGS__);							\
	}
#endif
```
But could not get it to work(still error C2059 at the same lines...).

**Comment by diz-vara** [comment]
aiworld, I've updated torch7 in my repo with latest changes - including your correction for THLapackCheck macro,

**Comment by dmandreev** [comment]
little bit hacky but not require to fix  https://github.com/aiworld/torch7/commit/4fe36336f918c235efaff98ef7469e248b4fce49#diff-6232529b35671ad635fed629c58a97f8

```
#if defined(_MSC_VER)
#define THLapackCheck(fmt, func, info , ...)						\
if (info < 0) {														\
  THError("Lapack Error in %s : Illegal Argument %d", func, -info); \
} else if(info > 0) {												\
  _THError(__FILE__,__LINE__, fmt, func, info, __VA_ARGS__);							\
}																	\


#else
#define THLapackCheck(fmt, func, info , ...)						\
if (info < 0) {														\
  THError("Lapack Error in %s : Illegal Argument %d", func, -info); \
} else if(info > 0) {												\
  THError(fmt, func, info, ##__VA_ARGS__);							\
}																	\

#endif
```
